### ISSUE TITLE
Chapter 3 Notes, Errata, and Discussion: Basic Scala

### ISSUE THREAD

**Issue by lihaoyi** [handsonscala/handsonscala#3]
Thread to discuss anything and everything about Chapter 3. Important topics will be folded into the issue description.

**Comment by EdwardFeng523** [comment]
The index access for Scala tuples is a bit strange compared with other languages. I'm just wondering why is that and is there any recommended library that we can use to convert it to normal index access (starting from 0)? Thanks

**Comment by lihaoyi** [comment]
@EdwardFeng523 that's just the way it is. It's definitely a bit unusual compared to what you may be used to in script languages, but you get used to it.

**Comment by ShapelessCat** [comment]
The title of 3.3.2 is "Function Values", and in 3.3.2.1 uses the term "function literal" without explaining the relation between it and "function values". I think this may confuse people who never saw these two concepts. I checked Martin's book, and Section 8.3 explains their relations in detail.
  Since chapter 3 is a part of the introduction of this book, I don't think explain the relation between "function values" and "function literals" is very important, but use one term consistently can help newcomers to understand.

**Comment by lihaoyi** [comment]
@ShapelessCat that's a good point. I guess the difference between *function value* and *function literal* is similar to *int value* and *int literal*: *values* are things you pass around, whereas *literals* are a syntax that creates a value you pass around (though you can also create *value* by other ways, e.g. the syntax `1 + 1` creates the value `2` even though there's no `2` literal in the code snippet). The two concepts are related, but not quite interchangeable

**Comment by ShapelessCat** [comment]
@lihaoyi Yes. Agree.
Here is what I found in the Programming in Scala, 4th edition:
- Section 8.3:
   > A _function literal_ is compiled into a class that when instantiated at runtime
is a _function value_. Thus the distinction between _function literals_ and
values is that _function literals_ exist in the source code, whereas _function values_
exist as objects at runtime. The distinction is much like that between
classes (source code) and objects (runtime).

- Glossary:
   + > **function literal** A function with no name in Scala source code, specified with function literal syntax. For example, (x: Int, y: Int) => x + y.

   + > **function value** A function object that can be invoked just like any other
function. A function value’s class extends one of the FunctionN traits
(e.g., Function0, Function1) from package scala, and is usually
expressed in source code via function literal syntax. A function value
is “invoked” when its apply method is called. A function value that
captures free variables is a closure.

I'm not sure what is the best way, but it seems according to theses definitions at least the first sentence in Section 3.3.2 Function Values should be "You can define function literals using the `=>` syntax."

**Comment by spatialy** [comment]
Hi @lihaoyi, I new to scala and FP.

I have a hard time understanding the solution for 3.5 - FlexibleFizzBuzz on the part for storing the results in the existing array ...

```
def flexibleFizzBuzz(handleLine: String => Unit): Unit = {
  for (i <- Range.inclusive(1, 100)) {
    handleLine(
      if (i % 3 == 0 && i % 5 == 0) "FizzBuzz"
      else if (i % 3 == 0) "Fizz"
      else if (i % 5 == 0) "Buzz"
      else i.toString
    )
  }
}

var i = 0
val output = new Array[String](100)

flexibleFizzBuzz{ s =>
  output(i) = s
  i += 1
}
```

Can you explain what happens when you pass the function ad parameter here? The receiving method has his internal counter, so the i += 1 is needed for?

Great book so far, by the way

Best

**Comment by implisci** [comment]
> I have a hard time understanding the solution for 3.5 - FlexibleFizzBuzz 

Rename the var i to iter, and change the i inside the function to iter.  This could help.  
```scala
var iter = 0
val output = new Array[String](100)

flexibleFizzBuzz{ s =>
  output(iter) = s
  iter += 1
  print(s"$iter ") // you should see it print from 1 to 100
                   // iter is not  i <- Range.inclusive(1, 100)
}
```
The flexible function can handle a variety of function parameters including the one you posted here.  When the flexibleFizzBuzz is called do you see the {} with the s => code ?  That is the function parameter handleLine which assigns the String s to the Array output (instead of printing it).  However the counter is defined outside the flexibleFunction.  When the function is passed it takes the iter variable (not to be confused with the i in the flexible Function) and has access to it even when it is inside the flexible function. Seems like magic but that's what happens.  So handleLine does not need and indeed cannot use the i inside the Range as all it can use is the String parameter (s) passed by the code inside handleLine which is an expression, not a statement. Expressions return values. 
```scala
if (i % 3 == 0 && i % 5 == 0) "FizzBuzz"
      else if (i % 3 == 0) "Fizz"
      else if (i % 5 == 0) "Buzz"
      else i.toString
```
Does this make sense? The above code evaluates as a String. handleLine is ignorant of how that came about (the logic using i etc). You are right that this is not easy to understand. But this concept is present in other languages too (JavaScript etc).  Here is a link on closures (that's what the concept is called):  
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures

**Comment by spatialy** [comment]
Hi @implisci, thanks for the help.

As you point if you change i for iter the code is better readable for newcomers like me.

Again thanks

best

**Comment by lihaoyi** [comment]
@spatialy Yes internal `i` counter within the loop and the external `i` counter being incremented by the function literal are different variables. The naming is certainly confusing and makes them sound like the same thing.

We keep different variables because the internal counter is not given to the `handleLine` callback, so the callback cannot make use of it. An alternative approach could be to pass the internal `i` to the `handleLine` callback to use:

```diff
-def flexibleFizzBuzz(handleLine: String => Unit): Unit = {
+def flexibleFizzBuzz(handleLine: (Int, String) => Unit): Unit = {
   for (i <- Range.inclusive(1, 100)) {
     handleLine(
+      i,
       if (i % 3 == 0 && i % 5 == 0) "FizzBuzz"
       else if (i % 3 == 0) "Fizz"
       else if (i % 5 == 0) "Buzz"
       else i.toString
     )
   }
 }

-var i = 0
 val output = new Array[String](100)

-flexibleFizzBuzz{ s =>
+flexibleFizzBuzz{ (i, s) =>
   output(i) = s
-  i += 1
 }
```

**Comment by jcon** [comment]
I enjoyed reading the first two chapters. I'm finding `amm` a pleasure to work in compared to my limited time with `sbt`'s REPL.

As an extension to the `flexibleFizzBuzz` exercise, I wanted to experiment hiding the counter using a class. I wanted to end up with something like below:

```scala
val acc = new Accumulator(100) 
flexibleFizzBuzz(acc.append) 
println(acc.buffer.mkString(",")) 
// -> 1,2,Fizz,4,Buzz,Fizz,...,98,Fizz,Buzz
```

Below is one implementation of `Accumulator` that satisfies above when we know the size of the input. 

```scala
class Accumulator(size: Int) {
    val buffer = new Array[String](size)
    private var position = 0

    def append(s: String): Unit = {
        buffer(position) = s
        position += 1
    }
}
```

I suspected I could replace `Accumulator` with a built-in vector type though and found `scala.collection.mutable.ArrayBuffer[String]` worked nicely:

```scala
val buf = new scala.collection.mutable.ArrayBuffer[String]() 
flexibleFizzBuzz(acc.append) 
println(buf.mkString(",")) 
// -> 1,2,Fizz,4,Buzz,Fizz,...,98,Fizz,Buzz
```

**Comment by HiCatherine** [comment]
In Chapter 3.5 Conclusion, the exercise asks to define a `flexibleFizzBuzz` function, but in the console that is showing the usages of this method, it says `fizzBuzz`. Might be a source of confusion for people :)

**Comment by omnipresent07** [comment]
"Define a `def flexibleFizzBuzz` method that takes a `String => Unit` callback function". To confirm, does the notion `String => Unit` means that the callback function takes `String` as input and returns `Unit` as output?

**Comment by omnipresent07** [comment]
> @spatialy Yes internal `i` counter within the loop and the external `i` counter being incremented by the function literal are different variables. The naming is certainly confusing and makes them sound like the same thing.
> 
> We keep different variables because the internal counter is not given to the `handleLine` callback, so the callback cannot make use of it. An alternative approach could be to pass the internal `i` to the `handleLine` callback to use:
> 
> ```diff
> -def flexibleFizzBuzz(handleLine: String => Unit): Unit = {
> +def flexibleFizzBuzz(handleLine: (Int, String) => Unit): Unit = {
>    for (i <- Range.inclusive(1, 100)) {
>      handleLine(
> +      i,
>        if (i % 3 == 0 && i % 5 == 0) "FizzBuzz"
>        else if (i % 3 == 0) "Fizz"
>        else if (i % 5 == 0) "Buzz"
>        else i.toString
>      )
>    }
>  }
> 
> -var i = 0
>  val output = new Array[String](100)
> 
> -flexibleFizzBuzz{ s =>
> +flexibleFizzBuzz{ (i, s) =>
>    output(i) = s
> -  i += 1
>  }
> ```

This will give IndexOutofBoundException unless we set the array size to be 101

**Comment by lihaoyi** [comment]
@omnipresent07 yes `String => Unit` means a function that takes a `String` and returns `Unit`. `Unit` is basically the same thing as `void` in other languages, meaning it doesn't need to return anything

**Comment by lihaoyi** [comment]
> This will give IndexOutofBoundException unless we set the array size to be 101

Yes, you are right. Or we set `output(i - 1) = s`

**Comment by Urzhumtsev1** [comment]
3.3.2.1 Methods taking Functions. Snippet 3.50
```
@ def increment(i: Int) = i + 1

@ val b = new Box(123)

@ b.update(increment) // Providing a method reference

@ b.update(x => increment(x)) // Explicitly writing out the lambda

@ b.update{x => increment(x)} // Methods that take a single lambda can be called with {}s

@ b.update(increment(_)) // You can also use the `_` placeholder syntax

@ b.printMsg("result: ")
result: 126
```
Why not 127?

**Comment by GitHCoRradO** [comment]
> 3.3.2.1 Methods taking Functions. Snippet 3.50
> 
> ```
> @ def increment(i: Int) = i + 1
> 
> @ val b = new Box(123)
> 
> @ b.update(increment) // Providing a method reference
> 
> @ b.update(x => increment(x)) // Explicitly writing out the lambda
> 
> @ b.update{x => increment(x)} // Methods that take a single lambda can be called with {}s
> 
> @ b.update(increment(_)) // You can also use the `_` placeholder syntax
> 
> @ b.printMsg("result: ")
> result: 126
> ```
> 
> Why not 127?

Should be 127.

**Comment by lihaoyi** [comment]
yeah should be 127, i'll update the next release

**Comment by colin-williams** [comment]
chapter 3

grammar: and use it to print out a threaded fashion. That means that child
messages are print out indented underneath their parents...

suggest (in a threaded fashion). Also run on sentence.

**Comment by colin-williams** [comment]
Regarding the `printMessages` exercise at the end for chapter 3. I'm not quite following how to recursively iterate through the array. The signature def

`def printMessages(messages: Array[Msg]): Unit = ...: Unit = ...`

doesn't provide an index. I could then define maybe

`def printMessages(messages: Array[Msg], index: Int): Unit=...`

but given no mention of recursion in the chapter it doesn't follow. I could just look at the solution... Finally should we expect the array elements are ordered, like provided?

**Comment by implisci** [comment]
@colin-williams 
 Yes, the Array of Messages is sorted by the Message ID in ascending order 0,1,2 etc and experience with writing recursive functions is assumed. The solution uses a common pattern in recursion. The outer function, the one that receives an Array of Messages as its only parameter calls a recursive helper function in its body which has 2 parameters, one of which is the parent ID (remember this is an Option, so if there is no parent then it is None). The approach is to kickstart the printing by calling this helper function which has 2 parameters, one of which is the parent ID and the other is the indent (this can be done in multiple ways as an Int or String) which is used for printing. The recursive part is to scan the entire array and filter the Messages that have the current parent ID as their parent and print that with a suitable indent. Having done that make a recursive call because we need to print the ones that might be children of the one that just got printed. As this is about printing no values are returned by the helper or outer function. In general, especially in the application chapters after chapter 6, it is ok to look at the solutions and make sense of them if you get stuck. But I would recommend trying to see if you can solve the exercises till the end of Chapter 6. Even there if you need to see the solution (this is ok), try to recreate it without seeing it after some time (when details are forgotten). Ultimately, it is about introspection and figuring out where the assumptions you've made don't fit the problem etc. and getting some aha moments. This problem teaches you how to think about and use Options within a recursive, effectful program and also to understand the order of printing as the recursive stack unwinds. Start with small examples and manually trace the execution.

**Comment by colin-williams** [comment]
Here's the naive solution I wrote:
```
 def printMessages(messages: Array[Msg]): Unit = { 
                 printMessages(messages, 0, 0, 0) 
                 def printMessages(messages: Array[Msg], index: Int, level: Int, lastIdx: Int): Unit = { 
                   if (index >= messages.length) {} 
                   else { 
                   messages(index) match { 
                     case Msg(id, None, txt) => { 
                       println(s"#$id,$txt") 
                       printMessages(messages,index+1, 0, id)} 
                     case Msg(id, Some(idx), txt) => { 
                       val lev = if (lastIdx == idx) level + 1 else level 
                       val levelString = "  " * lev 
                       println(s"${levelString}#$id,$txt") 
                       printMessages(messages,index+1,lev, id)} 
                   } 
                 } 
                 } 
 }
```

**Comment by implisci** [comment]
First thing. Don't worry about performance for now, O(n), O(n^2) etc. Aim to make the code simple and understandable.   Good to rename the helper function to something short like printOne or even just helper.  Less confusion.  The  printMessages call on top is ok but ideally it should be at the end of the function for better understanding (people look to the end to see return value).  messages is visible to the recursive function. It doesn't need to be passed as a parameter.  Can you combine both cases into one? Right now there are lots of parameters, 3 Ints.  Why not use the parent Option as a parameter?  Also you are checking indexes. Not recommended.  Ultimately if your code works it is fine but it is not maintainable as it is.  Here's how the official solution works.  helper has just 2 parameters.  the parent id as an option Some(id) is an option and an indent string.  You can try this for some time and then examine the solution and ponder why it is so much cleaner.  Again we are not focusing on performance here.  There are ways to get there as well but thats not the key idea for this exercise. Are you using backticks with scala to get code formatting here using markdown? three backticks with the word scala following and three backticks after the code. Ideally indent the code properly from your coding editor (visual studio code has a way to format code)

**Comment by georgeteo** [comment]
Is it common in scala code to elide the `()` for 0 params method calls and is there a way to invoke a no-args function without executing it? 

```
@ def g(s: String): Unit = { println(s"g($s)") }
defined function g

// thanks for the helpful error message, so we need to type variable assignment of functions
@ val gValUntyped = g
cmd55.sc:1: missing argument list for method g in object cmd51
Unapplied methods are only converted to functions when a function type is expected.
You can make this conversion explicit by writing `g _` or `g(_)` instead of `g`.
val gValUntyped = g
              ^
Compilation Failed
 
@ val gValTyped : String => Unit = g
gVal: String => Unit = ammonite.$sess.cmd52$$$Lambda$1913/804949221@12e4c6de

@ gValTyped("some str")
g(some str)

@ def f() : Unit = { println("f()") }
defined function f

// lets try untyped variable assignment for Unit => Unit method.
@ val fValUntyped = f
f()
// it worked without error message?

// oh actually, it didn't work since scala implicitly executed the method f without args
// and returned unit.
@ fValUntyped("some str")
cmd56.sc:1: Unit does not take parameters
val res56 = fVal("some str")
                ^
Compilation Failed

// But we learned previously that variable assignment from functions should be typed,
// maybe that will help.
@ val fValTyped : Unit => Unit = f
cmd56.sc:1: type mismatch;
 found   : Unit
 required: Unit => Unit
val fValTyped : Unit => Unit = f
                               ^
Compilation Failed

// nope, didn't help. 
```

As an aside, It feels inconsistent and results in difficult (for newbies) to comprehend behavior.

**Comment by implisci** [comment]
@colin-williams The printMessages code that you posted requires Msg be a case class (for the matching to work). The official book solution will work even if the Msg entries are reversed or shuffled (the display of the threads will be different but essentially correct).  If the input Array has a .sortBy(_.id) then the display will be identical. Regarding Arrays and Indices: Yes, they are a powerful feature and indispensable often (Sudoku example). But sometimes they are not needed. I'm not sure if your code above will display threads correctly in all cases. See the following input array for example. The second last entry has a different parent. Compare your output to the official code. Notice that the outputs are different, suggesting that your approach needs review.
```scala 
printMessages(Array(
                 new Msg(1, Some(0), "World"),
                 new Msg(0, None, "Hello"),
                 new Msg(2, None, "I am Cow"),
                 new Msg(3, Some(2), "Hear me moo"),
                 new Msg(4, Some(2), "Here I stand"),
                 new Msg(5, Some(0), "I am Cow"), // changed parent
                 new Msg(6, Some(5), "Here me moo, moo")
               )) 
```

**Comment by implisci** [comment]
@georgeteo 
The meaning of () is context specific, like the underscore in Scala. () could  be the singleton whose type is Unit and indicates the absence of a meaningful value. () could also represent a method or function that takes no parameters. The defs are methods and when they are lifted they become functions.  There are some practical distinctions between methods and functions (set to almost disappear in Scala 3) such as that functions can't have default values for parameters or named parameters but methods can have both. The syntax for lifting a method into a function is to place an _ after the function name (no need to specify the type). There are multiple stackoverflow threads about lifting. Functions and methods can have multiple parameter lists. This makes possible "currying" (creating more specific functions from more general functions). See https://alvinalexander.com/scala/fp-book/partially-applied-functions-currying-in-scala/  Your last example in the REPL shows that the type is () => Unit where () means no parameters (not the Unit singleton). If you had defined f as def f(x:Unit):Unit, it would have worked as below.  For methods it is common to use empty () if the method is run for a side effect like println.  This provides a visual cue to the side-effecty nature of the method. If the method returns a value and does not have significant side effects, the () is often dropped. Functions can have 0 to 22 parameters with corresponding types Function0 to Function22.
```scala
def f(x:Unit) = println("foo") 
// defined function f
val f1 = f _ 
// f1: Unit => Unit 
f1(())
// foo
def g() = println("bar")
// defined function g
val g1 = g _ 
// g1: () => Unit  Here () stands for a method or function with no parameters
g1()
// bar
// Aside on Nothing
def z = ???  
// def z:Nothing
// no compile error
z
// scala.NotImplementedError: an implementation is missing
```
Additional Info: The type Nothing has no instances/values and is a subtype of every type in Scala. Nothing is a bottom class in the Scala type hierarchy. The Unit type has one instance () and indicates lack of meaningful value.  None has type Option[Nothing], Nil has type List[Nothing], and null has type Null which is a subtype of everything except AnyVals like Int. In scala, null cannot be assigned to value classes such as Int. So the type Null is also a bottom class but not as comprehensive as Nothing in its bottomness.  A note on Tuples:  Empty Tuples are not possible or definable in Scala, unlike Python. Tuples have types ranging from Tuple1 to Tuple22 with a specification of types inside [ ] following the TupleXX. Collections can be empty but not Tuples. So () does not represent an empty Tuple (which does not exist) in Scala.
![image](https://user-images.githubusercontent.com/4403773/88849677-387b1700-d1b8-11ea-9569-2956dc6ef5ef.png)

**Comment by lihaoyi** [comment]
@georgeteo optional `()`s are definitely confusing; I myself asked this stackoverflow question many years ago on the topic https://stackoverflow.com/questions/8303817/nine-ways-to-define-a-method-in-scala. There are a number of good answers that explain how the mechanics of optional `()`s work.

On the bright side, optional `()`s are being removed in Scala 3, in favor of always having to follow the definition site whether it uses `()` or not. So this point of confusion should go away in future, and for now the best convention is follow the definition of a method whether or not it uses `()`

**Comment by Bolshem** [comment]
@lihaoyi Hi

Just a small comment about exercise.

You are asking to write a recursive method called printMessages and you give the definition of this method.
`def printMessages(messages: Array[Msg]): Unit = ...`

 By the definition, recursive method is the method that calls itself, and even in your solution, the real recursive method is the INTERNAL method 

`def printFrag(parent: Option[Int], indent: String): Unit =...`

and not the method printMessages. Method "printMessages" doesn't call itself, thus it is not recursive. It uses the internal recursive method. So your solution is not exactly what you ask to do :)

If you want printMessages to be recursive, then the solution will be a little ugly with external stacks (to preserve parents/levels)

`
`

**Comment by mirandaio** [comment]
Hi @lihaoyi 

Is example 3.51.scala idiomatic Scala or just a matter of taste? Is it preferred to define the `myLoop` method like

```scala
def myLoop(start: Int, end: Int) (callback: Int => Unit) = {
  for (i <- Range(start, end)) {
    callback(i)
  }
}
```

instead of

```scala
def myLoop(start: Int, end: Int, callback: Int => Unit) = {
  for (i <- Range(start, end)) {
    callback(i)
  }
}
```

I find the second way to define the method easier to read.

Also, when invoking the method, I find

```scala
myLoop(start = 5, end = 10, i => println(s"i has value ${i}"))
```

easier to read than

```scala
myLoop(start = 5, end = 10) { i =>
  println(s"i has value ${i}")
}
```

Thanks!

**Comment by mickeyvip** [comment]
Here is my implementation of `printMessages` (some hints are taken from this thread):

```scala
  def printMessages(messages: Array[Msg]): Unit = {
    val rootParents = messages.filter(_.parent.isEmpty)

    def innerPrint(innerMessages: Array[Msg], level: Int = 0): Unit = {
      for (msg <- innerMessages) {
        println(("  " * level) + s"#${msg.id} ${msg.txt}")
        
        val children = messages.filter(_.parent.contains(msg.id))
        innerPrint(children, level + 1)
      }
    }

    innerPrint(rootParents)
  }
```

First we take all the "root" messages and then start the recursion by calling `innerPrint` with them and nesting level `0`. The `innerPrint` loops over the passed in `innerMessages` and for each it:
- prints the message indenting it as needed
- finds all the messages who's parent is the current message
- calls the `innerPrint` with them and increases the nesting level

**Comment by Gemuruh-Geo** [comment]
Here is my solution of printMessages, I cannot wrap my head to write solution in single loop, so I am divide my thinking process into two step

1) first get ultimate parent (all parent with parent id None)
2) get all children, potentially also a parent 

this solution also work, if we shuffle Array Message Element
 
```
def printMessages(arrayMsg: Array[Msg]): Unit = {

    val allUltimateParent = arrayMsg.filter(msg => msg.parent.isEmpty) //All Parent with parent id none
    val allChildren = arrayMsg.diff(allUltimateParent) //All children that not ultimate parent but, its maybe also a parent
    
    allUltimateParent.foreach(m => {
      println(s"#${m.id} ${m.txt}")
      h(m,allChildren,1)
    })

    def h(msg: Msg, array:Array[Msg], k: Int): Unit = {
      if(array.isEmpty)return

      val id = msg.id
      val children = array.filter(msg => {
        val parentId = msg.parent.getOrElse(-1)
        id == parentId
      })
      children.foreach(ch => {
        println(("\t"*k)+s"#${ch.id} ${ch.txt}")
        h(ch, array.diff(Seq(ch)),k+1)
      })

    }

  }


```

**Comment by MBrighty** [comment]
Am I right in thinking that the test in [TestContextManagers.sc](https://github.com/handsonscala/handsonscala/blob/v1/examples/3.7%20-%20ContextManagers/TestContextManagers.sc#L4-L7) is always going to fail with the example solution given [here](https://github.com/handsonscala/handsonscala/blob/v1/examples/3.7%20-%20ContextManagers/ContextManagers.sc)? `withFileReader` doesn't actually return anything and the value of `result` will be `Unit` causing the assertion to fail.
With the example given I had to assign result in the callback passed to `withFileReader` like so:
```scala
var result = ""

withFileReader("File.txt") { reader => 
  result = reader.readLine() + "\n" + reader.readLine()
}

assert(result == "Hello\nWorld!")
```

**Comment by pagpires** [comment]
> Am I right in thinking that the test in [TestContextManagers.sc](https://github.com/handsonscala/handsonscala/blob/v1/examples/3.7%20-%20ContextManagers/TestContextManagers.sc#L4-L7) is always going to fail with the example solution given [here](https://github.com/handsonscala/handsonscala/blob/v1/examples/3.7%20-%20ContextManagers/ContextManagers.sc)? `withFileReader` doesn't actually return anything and the value of `result` will be `Unit` causing the assertion to fail.
> With the example given I had to assign result in the callback passed to `withFileReader` like so:
> 
> ```scala
> var result = ""
> 
> withFileReader("File.txt") { reader => 
>   result = reader.readLine() + "\n" + reader.readLine()
> }
> 
> assert(result == "Hello\nWorld!")
> ```

The solution did run ok: double check whether you have the generic typing `[T]` for withFileReader, which actually returns a `String` in this case and `T` cannot be omitted, although it can be for `withFileWriter` (i.e. changed to `Unit`)

**Comment by runnerlt** [comment]
Hi, 
Exercise 3.7 - ContextManagers uses syntax, which is not described in the previous chapters( **=> [T]** ).
`def withFileWriter[T](fileName: String)(handler: java.io.BufferedWriter => T)`

I expected this book to be consistent when buying it. I'm newbie in Scala and FP. And things like these demotivates a lot.

**Comment by rahulj51** [comment]
Hi,

There is a very minor grammatical error in the beginning paragraph.

`The goal of this chapter is to get familiar you enough` should read as

`The goal of this chapter is to get you familiar enough`

**Comment by jsatk** [comment]
I found exercise 3.7 to be a bit much. It requires you digging into the java standard library API and learning, on your own, how to figure out how to read and write files. This isn't covered at all in the previous chapter. This is quite a leap from learning the basics like `trait`, `val`, etc. You might want to include the methods a user would need to read and write files.

**Comment by JayStringer** [comment]
Hey, enjoying the book so far, but with regards to problem 3.6, I don't think it's fair to assume experience writing recursive functions. For future iterations of the book I think it might be a good idea to either:

- Have a quick primer on recursion in  chapter 3
- Have a short section in chapter 1 discussing "Who is this book for?"
- Have a short section in chapter 1 discussing the assumptions that this book makes with regards to experience of the reader

Thanks

**Comment by lihaoyi** [comment]
@jsatk @JayStringer I think both of these concerns basically boil down to the target audience:

> Hands-on Scala assumes you are a software developer who already has experience working in another programming language, and want to quickly become productive working with Scala.

> Note that this book is not targeted at complete newcomers to programming. We expect you to be familiar with basic programming concepts: variables, integers, conditionals, loops, functions, classes, and so on.

> We will touch on any cases where these concepts behave differently in Scala, but expect that you already have a basic understanding of how they work.

While it's arguable what "a software developer" means, being able to figure out reading/writing to files using a (possible unfamiliar) standard library or writing a simple recursive function are certainly reasonable expectations. If nothing else, both of these are things I would expect someone to learn pretty early in any N-year CS degree, and many employers certainly expect you to be able to do these things on-the-spot during a 60-minute job interview (for better or worse).

Some of these exercises are certainly not trivial, but that's why they're exercises: they're meant to make you work and break a sweat! 

Finally, the book's exercises have their solutions available online specifically for you to learn from, so don't feel bad about looking at them. Any professional software developer should be able to study existing code to learn techniques they couldn't figure out on their own. That's not a failure on anyone's part, that's just the reality of the work we do every day.

**Comment by lihaoyi** [comment]
@runnerlt yeah the syntax for defining generic functions is certainly something I missed covering early enough. The exercise is loosely specified enough that you could have written a non-generic version hard-coded to `String` though. Regardless, I hope the online example was useful.

**Comment by JayStringer** [comment]
@lihaoyi Thanks for taking the time to respond

I apologise, I completely forgot about the quoted parts of the book you shared which do address my latter two points. And yes I do agree with you for the most part.

Perhaps I am in a minority of working engineers who haven't done a CS degree. As such many of the year one CS fundamentals that undergrads learn, like recursion and even basic algorithms for example (which haven't as yet been required for me to learn; they just haven't been needed for the work I've done), are gaps in my knowledge portfolio.

Anyway, as I mentioned I'm likely in the minority of readers and perhaps not the target audience. Regardless, I am enjoying myself so far, cheers :)

**Comment by jsatk** [comment]
> @jsatk @JayStringer I think both of these concerns basically boil down to the target audience:
> 
> > Hands-on Scala assumes you are a software developer who already has experience working in another programming language, and want to quickly become productive working with Scala.
> 
> > Note that this book is not targeted at complete newcomers to programming. We expect you to be familiar with basic programming concepts: variables, integers, conditionals, loops, functions, classes, and so on.
> 
> > We will touch on any cases where these concepts behave differently in Scala, but expect that you already have a basic understanding of how they work.
> 
> While it's arguable what "a software developer" means, being able to figure out reading/writing to files using a (possible unfamiliar) standard library or writing a simple recursive function are certainly reasonable expectations. If nothing else, both of these are things I would expect someone to learn pretty early in any N-year CS degree, and many employers certainly expect you to be able to do these things on-the-spot during a 60-minute job interview (for better or worse).
> 
> Some of these exercises are certainly not trivial, but that's why they're exercises: they're meant to make you work and break a sweat!
> 
> Finally, the book's exercises have their solutions available online specifically for you to learn from, so don't feel bad about looking at them. Any professional software developer should be able to study existing code to learn techniques they couldn't figure out on their own. That's not a failure on anyone's part, that's just the reality of the work we do every day.

Thanks for the kind response. And I agree with you. I was just stunned to see exercises asking me to do things that were outside of what was taught in the chapter. I've been programming for 11+ years and doing scala for 2y. I was able to figure out the recursion one but the read/write I had to look up your example. It's simply something I've never had to touch in scala or java.

Thanks for clarifying that the exercises expect a little more than what was just in the chapter.

**Comment by mferrall** [comment]
In the solution, for ContextWriters, is the choice to do the curried format (I think its curried, in that you have separate parens for each parameter) just a stylistic choice?  Is that idiomatic in scala when you have a mix of values and fuctions?
```scala
def withFileWriter[T](fileName: String)(handler: java.io.BufferedWriter => T) = {
  val output = java.nio.file.Files.newBufferedWriter(java.nio.file.Paths.get(fileName))
  try handler(output)
  finally output.close()
}

**Comment by cdcd88** [comment]
@lihaoyi  What is `[T]` in exercise 3.7 solution?

```
def withFileWriter[T](fileName: String)(handler: java.io.BufferedWriter => T) = {
  val output = java.nio.file.Files.newBufferedWriter(java.nio.file.Paths.get(fileName))
  try handler(output)
  finally output.close()
}
```

The method that calls it doesn't specify anything about `[T]` at all.

```
withFileWriter("File.txt") { writer =>
  writer.write("Hello\n"); writer.write("World!")
}
```

**Comment by JayStringer** [comment]
@cdcd88 Hey, `T` Is a generic type parameter. The second argument list for `withFileWriter` has an argument called `handler`, which should be a function that takes `java.io.BufferedWriter` and returns whatever type `T` is specified to be.

For example, `withFileWriter[String]` requires a `handler` function from `java.io.BufferedWriter` to `String`

Hope that helps :)

**Comment by cdcd88** [comment]
> @cdcd88 Hey, `T` Is a generic type parameter. The second argument list for `withFileWriter` has an argument called `handler`, which should be a function that takes `java.io.BufferedWriter` and returns whatever type `T` is specified to be.
> 
> For example, `withFileWriter[String]` requires a `handler` function from `java.io.BufferedWriter` to `String`
> 
> Hope that helps :)

Does the compiler infer `[T]` from code?

For example, the block below works (`[T]` is `String`)
```
withFileWriter("File.txt") { writer =>
  writer.write("Hello\n"); writer.write("World!")
  "test"
}
```

This block of code also works (`[T]` is `Unit`)
```
withFileWriter("File.txt") { writer =>
  writer.write("Hello\n"); writer.write("World!")
}
```

However, if I specify the generic parameter, the compiler enforces the return type of that caller. The code below doesn't work because I specified `[String]` while the function returns `Unit`
```
withFileWriter[String]("File.txt") { writer =>
  writer.write("Hello\n"); writer.write("World!")
}
```

**Comment by cdcd88** [comment]
I have two questions about naming convention and code format. I wrote the following code for exercise 4.16 

```
def renderSudoku4(grid: Array[Array[Int]]) = {
  val separator = "+-------+-------+-------+\n"
  grid.map(rows => rows.map(value => if (value == 0) " " else value.toString).grouped(3).map(_.mkString(" ")).mkString("| ", " | ", " |")).grouped(3).map(_.mkString(separator, "\n", "\n")).mkString + separator
}
```

First, I named `rows` in the first map function. I have a background in .NET, and its convention is to name explicit and meaningful variable names in code.

Second, this line is long. Again, in .NET, I would add a line break after `=>` since it introduces a new logical group. I found https://docs.scala-lang.org/style/indentation.html#line-wrapping providing a general guideline, but I am not entirely clear if I should focus on 80-character length or there are other recommendations. 

I wanted to know where I can read more about these two topics. Please advise.

**Comment by lihaoyi** [comment]
@cdcd88 it's really a matter of preference. I don't think it's really a strict rule in .NET either; when using LINQ-style method-chaining APIs like `.Where` and `.Select`, it's just as common to chain a bunch of them together without intermediate variables. In fact, often intermediate variables aren't even possible, e.g. when dealing with LINQ-to-SQL or similar transformations

Generally naming intermediate variables is entirely up to you, to do so when it helps and not do so when it doesn't. As for line-splitting/indentation/whitespace, the `scalafmt` defaults are a pretty good place to start (which is great, because you don't need to deal with it manually!)

**Comment by bergamobobson** [comment]
My solution to printMessage. The solution provide with the book was not quite intuitive for me.
def printMessages(messages: Array[Msg]): Unit = {
    if (messages.nonEmpty) {
      val msg = messages.head
      val indentation = " " * msg.parent.map(_ + 1).getOrElse(0)
      println(indentation + msg.txt)
      printMessages(messages.tail)
    }
  }
